pkg/instancetype/instancetype.go

```diff
@@ -521,6 +531,123 @@ func (m *methods) findClusterInstancetype(vm *virtv1.VirtualMachine) (*instancet
+		}
+		if volume.DataVolume != nil {
+			return m.inferDefaultsFromDataVolume(vm, volume.DataVolume.Name, defaultNameAnnotation, defaultKindAnnoation)
+		}
+		return "", "", fmt.Errorf("unable to infer defaults from Volume as type is not supported")
+	}
+	return "", "", fmt.Errorf("unable to find volume %s to infer defaults", inferFromVolumeName)
+}
+
+func inferDefaultsFromAnnotations(annotations map[string]string, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	defaultName, hasAnnotation := annotations[defaultNameAnnotation]
+	if !hasAnnotation {
+		return "", "", fmt.Errorf("unable to find required %s annotation on the volume", defaultNameAnnotation)
+	}
+	return defaultName, annotations[defaultKindAnnoation], nil
+}
+
+func (m *methods) inferDefaultsFromPVC(pvcName, pvcNamespace, defaultNameAnnotation, defaultKindAnnotation string) (string, string, error) {
+	pvc, err := m.clientset.CoreV1().PersistentVolumeClaims(pvcNamespace).Get(context.Background(), pvcName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", err
+	}
+	return inferDefaultsFromAnnotations(pvc.Annotations, defaultNameAnnotation, defaultKindAnnotation)
+}
+
+func (m *methods) inferDefaultsFromDataVolume(vm *virtv1.VirtualMachine, dvName, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	if len(vm.Spec.DataVolumeTemplates) > 0 {
+		for _, dvt := range vm.Spec.DataVolumeTemplates {
+			if dvt.Name != dvName {
+				continue
+			}
+			return m.inferDefaultsFromDataVolumeSpec(dvt.Spec, defaultNameAnnotation, defaultKindAnnoation)
+		}
+	}
+	dv, err := m.clientset.CdiClient().CdiV1beta1().DataVolumes(vm.Namespace).Get(context.Background(), dvName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", err
+	}
+	// Check the DataVolume for any annotations before checking the underlying PVC
+	defaultName, defaultKind, err := inferDefaultsFromAnnotations(dv.Annotations, defaultNameAnnotation, defaultKindAnnoation)
+	if err == nil {
+		return defaultName, defaultKind, nil
+	}
+	return m.inferDefaultsFromDataVolumeSpec(dv.Spec, defaultNameAnnotation, defaultKindAnnoation)
+}
+
+func (m *methods) inferDefaultsFromDataVolumeSpec(dataVolumeSpec v1beta1.DataVolumeSpec, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	if dataVolumeSpec.Source != nil && dataVolumeSpec.Source.PVC != nil {
+		return m.inferDefaultsFromPVC(dataVolumeSpec.Source.PVC.Name, dataVolumeSpec.Source.PVC.Namespace, defaultNameAnnotation, defaultKindAnnoation)
+	}
+	if dataVolumeSpec.SourceRef != nil {
+		return m.inferDefaultsFromDataVolumeSourceRef(dataVolumeSpec.SourceRef, defaultNameAnnotation, defaultKindAnnoation)
+	}
+	return "", "", fmt.Errorf("unable to infer defaults from DataVolumeSpec as DataVolumeSource is not supported")
+}
+
+func (m *methods) inferDefaultsFromDataSource(dataSourceName, dataSourceNamespace, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
```

Comment:
Order of methods: `inferDefaultsFromDataVolumeSourceRef` gets referred to before `inferDefaultsFromDataSource`